### PR TITLE
Ota/ota demo core http

### PR DESCRIPTION
**Pull request by pvyawaha**
*Issue #, if available:*

*Description of changes:*
The change adds OTA over HTTP demo and  configurations files along with utility files for url parsing.

By submitting this pull request, I confirm that my contribution is made under the terms of the MIT license.

### CODE REVIEW

File: demos/ota/ota_demo_core_http/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+# Demo target.
+add_executable(${DEMO_NAME})
+
+# librt is required to use OTA OS POSIX port.
+find_library(LIB_RT rt REQUIRED)
+
+target_sources(
+    ${DEMO_NAME}
+    PRIVATE
+        "ota_demo_core_http.c"
```
Comment (aggarw13) [inline demos/ota/ota_demo_core_http/CMakeLists.txt:19]:
```suggestion
        "${DEMO_NAME}.c"
```

File: demos/ota/ota_demo_core_http/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        "${DEMOS_DIR}/ota/common/src/http_demo_utils.c"
+        ${OTA_SOURCES}
+        ${OTA_OS_POSIX_SOURCES}
+        ${OTA_MQTT_SOURCES}
+        ${OTA_HTTP_SOURCES}
+        ${MQTT_SOURCES}
+        ${MQTT_SERIALIZER_SOURCES}
+        ${HTTP_SOURCES}
+        ${BACKOFF_ALGORITHM_SOURCES}
+)
```
Comment (aggarw13) [inline demos/ota/ota_demo_core_http/CMakeLists.txt:30]:
```suggestion
)

```

File: demos/ota/ota_demo_core_http/demo_config.h [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+ * 2. Define LIBRARY_LOG_NAME and  LIBRARY_LOG_LEVEL.
+ * 3. Include the header file "logging_stack.h".
+ */
+
+/* Include header that defines log levels. */
+#include "logging_levels.h"
+
+/* Logging configuration for the Demo. */
+#ifndef LIBRARY_LOG_NAME
+    #define LIBRARY_LOG_NAME     "DEMO"
```
Comment (aggarw13) [inline demos/ota/ota_demo_core_http/demo_config.h:40]:
```suggestion
    #define LIBRARY_LOG_NAME     "OTADEMO"
```

File: demos/ota/ota_demo_core_http/demo_config.h [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+/**
+ * @brief Details of the MQTT broker to connect to.
+ *
+ * @note Your AWS IoT Core endpoint can be found in the AWS IoT console under
+ * Settings/Custom Endpoint, or using the describe-endpoint API.
+ *
+ * #define AWS_IOT_ENDPOINT               "...insert here..."
+ */
+#define AWS_IOT_ENDPOINT    "a3e1obletr43ue-ats.iot.us-west-2.amazonaws.com"
```
Comment (aggarw13) [inline demos/ota/ota_demo_core_http/demo_config.h:58]: A valid endpoint should not be provided!
```suggestion
```

File: demos/ota/ota_demo_core_http/demo_config.h [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+ *
+ * @note This certificate should be PEM-encoded.
+ * @note This path is relative from the demo binary created. Update
+ * ROOT_CA_CERT_PATH to the absolute path if this demo is executed from elsewhere.
+ */
+#ifndef ROOT_CA_CERT_PATH
+    #define ROOT_CA_CERT_PATH    "certificates/AmazonRootCA1.crt"
+#endif
+
+#ifndef ROOT_CA_CERT_PATH_HTTP
```
Comment (aggarw13) [inline demos/ota/ota_demo_core_http/demo_config.h:102]: There should be documentation about why there is a separate root CA for HTTP connection. 
Also, the root cert used for MQTT connection can be renamed to `AWS_IOT_ROOT_CA_CERT_PATH` to make it clear that it represents the root CA cert for AWS IoT.

File: demos/ota/ota_demo_core_http/demo_config.h [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+ *
+ * Refer to the AWS documentation below for details regarding client
+ * authentication.
+ * https://docs.aws.amazon.com/iot/latest/developerguide/client-authentication.html
+ *
+ * @note This certificate should be PEM-encoded.
+ *
+ * #define CLIENT_CERT_PATH    "...insert here..."
+ */
+#define CLIENT_CERT_PATH           "/home/risc/work/iot/things/otalinux/cert.crt"
```
Comment (aggarw13) [inline demos/ota/ota_demo_core_http/demo_config.h:117]:
```suggestion
```

File: demos/ota/ota_demo_core_http/demo_config.h [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+ *
+ * Refer to the AWS documentation below for details regarding client
+ * authentication.
+ * https://docs.aws.amazon.com/iot/latest/developerguide/client-authentication.html
+ *
+ * @note This private key should be PEM-encoded.
+ *
+ * #define CLIENT_PRIVATE_KEY_PATH    "...insert here..."
+ */
+#define CLIENT_PRIVATE_KEY_PATH    "/home/risc/work/iot/things/otalinux/pri.key"
```
Comment (aggarw13) [inline demos/ota/ota_demo_core_http/demo_config.h:130]:
```suggestion
```

File: demos/ota/ota_demo_core_http/demo_config.h [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+
+/**
+ * @brief MQTT client identifier.
+ *
+ * No two clients may use the same client identifier simultaneously.
+ */
+#ifndef CLIENT_IDENTIFIER
+    #define CLIENT_IDENTIFIER    "praz-thing-8-18-20"
```
Comment (aggarw13) [inline demos/ota/ota_demo_core_http/demo_config.h:138]: Why such a unique default string for the client ID?

File: demos/ota/ota_demo_core_http/demo_config.h [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+ * #define CLIENT_PRIVATE_KEY_PATH    "...insert here..."
+ */
+#define CLIENT_PRIVATE_KEY_PATH    "/home/risc/work/iot/things/otalinux/pri.key"
+
+/**
+ * @brief MQTT client identifier.
+ *
+ * No two clients may use the same client identifier simultaneously.
+ */
+#ifndef CLIENT_IDENTIFIER
```
Comment (aggarw13) [inline demos/ota/ota_demo_core_http/demo_config.h:137]: Suggest using the `MQTT` prefix to allow it to be distinguished from the HTTP connection used by the demo.
```suggestion
#ifndef MQTT_CLIENT_IDENTIFIER
```

File: demos/ota/ota_demo_core_http/ota_config.h [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+ */
+#define otaconfigLOG2_FILE_BLOCK_SIZE    10UL
+
+/************ End of logging configuration ****************/
+
+/**
+ * @brief Size of the file data block message (excluding the header).
+ *
+ */
+#define otaconfigFILE_BLOCK_SIZE               1024UL
```
Comment (aggarw13) [inline demos/ota/ota_demo_core_http/ota_config.h:69]:
```suggestion
#define otaconfigFILE_BLOCK_SIZE               ( 1UL << otaconfigLOG2_FILE_BLOCK_SIZE )
```

File: demos/ota/ota_demo_core_http/ota_config.h [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+/************ End of logging configuration ****************/
+
+/**
+ * @brief Log base 2 of the size of the file data block message (excluding the header).
+ *
+ * 10 bits yields a data block size of 1KB.
+ */
+#define otaconfigLOG2_FILE_BLOCK_SIZE    10UL
+
+/************ End of logging configuration ****************/
```
Comment (aggarw13) [inline demos/ota/ota_demo_core_http/ota_config.h:63]:
```suggestion
```

File: demos/ota/ota_demo_core_http/ota_config.h [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+ * the OTA agent for job and file data blocks received.
+ */
+#define otaconfigMAX_NUM_OTA_DATA_BUFFERS      1U
+
+/**
+ * @brief Allow update to same or lower version.
+ *
+ * Set this to 1 to allow downgrade or same version update.This configurations parameter
+ * disables version check and allows update to a same or lower version.This is provided for
+ * testing purpose and it is recommended to always update to higher version and keep this
```
Comment (aggarw13) [inline demos/ota/ota_demo_core_http/ota_config.h:130]:
```suggestion
 * Set this to 1 to allow version update to a lower or the same version. This configurations parameter
 * disables version check and allows update to a same or lower version. This is provided for
 * testing purpose. It is RECOMMENDED to always update to higher version and keep this
```

File: demos/ota/ota_demo_core_http/ota_demo_core_http.c [hunk 13]
```diff
@@ -0,0 +1,1620 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file ota_demo_core_mqtt.c
```
Comment (aggarw13) [inline demos/ota/ota_demo_core_http/ota_demo_core_http.c:23]:
```suggestion
 * @file ota_demo_core_http.c
```

File: demos/ota/ota_demo_core_http/ota_demo_core_http.c [hunk 14]
```diff
@@ -0,0 +1,1620 @@
...
+#include "ota_appversion32.h"
+
+/**
+ * These configuration settings are required to run the OTA demo which uses mutual authentication.
+ * Throw compilation error if the below configs are not defined.
+ */
+#ifndef AWS_IOT_ENDPOINT
+    #error "Please define AWS IoT MQTT broker endpoint(AWS_IOT_ENDPOINT) in demo_config.h."
+#endif
+#ifndef ROOT_CA_CERT_PATH
```
Comment (aggarw13) [inline demos/ota/ota_demo_core_http/ota_demo_core_http.c:78]: The check for the HTTP root CA cert also needs to be added

File: demos/ota/ota_demo_core_http/ota_demo_core_http.c [hunk 15]
```diff
@@ -0,0 +1,1620 @@
...
+};
+
+/* Each compilation unit must define the NetworkContext struct. */
+struct NetworkContext
+{
+    OpensslParams_t * pParams;
+};
+
+/**
+ * @brief Network connection context used in this demo.
```
Comment (aggarw13) [inline demos/ota/ota_demo_core_http/ota_demo_core_http.c:247]:
```suggestion
 * @brief Network connection context used in this demo for MQTT connection.
```

File: demos/ota/ota_demo_core_http/ota_demo_core_http.c [hunk 16]
```diff
@@ -0,0 +1,1620 @@
...
+ */
+static NetworkContext_t networkContext;
```
Comment (aggarw13) [inline demos/ota/ota_demo_core_http/ota_demo_core_http.c:249]:
```suggestion
static NetworkContext_t networkContextMqtt;
```

File: demos/ota/ota_demo_core_http/ota_demo_core_http.c [hunk 17]
```diff
@@ -0,0 +1,1620 @@
...
+
+/**
+ * @brief Structure for openssl parameters.
+ */
+static OpensslParams_t opensslParams;
+
+/**
+ * @brief Structure for openssl parameters.
+ */
+static OpensslParams_t opensslParamsHttp;
```
Comment (aggarw13) [inline demos/ota/ota_demo_core_http/ota_demo_core_http.c:296]:
```suggestion
/**
 * @brief Structure for openssl parameters for TLS session used by MQTT connection.
 */
static OpensslParams_t opensslParamsForMqtt;

/**
 * @brief Structure for openssl parameters for TLS session used by HTTP connection.
 */
static OpensslParams_t opensslParamsForHttp;
```

**Review by n9wxu — DISMISSED** [review]

File: demos/ota/ota_demo_core_http/ota_demo_core_http.c (hunk 17 above)
Comment (divekarshubham) [reply demos/ota/ota_demo_core_http/ota_demo_core_http.c]: Addressed in https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1425/commits/9f18e9905580d6e4067eaa4a215da9dbe1e2453a

File: demos/ota/ota_demo_core_http/ota_demo_core_http.c (hunk 16 above)
Comment (divekarshubham) [reply demos/ota/ota_demo_core_http/ota_demo_core_http.c]: Addressed in https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1425/commits/9f18e9905580d6e4067eaa4a215da9dbe1e2453a

File: demos/ota/ota_demo_core_http/ota_demo_core_http.c (hunk 14 above)
Comment (divekarshubham) [reply demos/ota/ota_demo_core_http/ota_demo_core_http.c:78]: Addressed in https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1425/commits/9f18e9905580d6e4067eaa4a215da9dbe1e2453a

File: demos/ota/ota_demo_core_http/demo_config.h (hunk 5 above)
Comment (divekarshubham) [reply demos/ota/ota_demo_core_http/demo_config.h:111]: Addressed in https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1425/commits/9f18e9905580d6e4067eaa4a215da9dbe1e2453a

File: demos/ota/ota_demo_core_http/CMakeLists.txt (hunk 2 above)
Comment (divekarshubham) [reply demos/ota/ota_demo_core_http/CMakeLists.txt:25]: Changed cmake in  https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1425/commits/9f18e9905580d6e4067eaa4a215da9dbe1e2453a

File: demos/ota/ota_demo_core_http/CMakeLists.txt (hunk 1 above)
Comment (divekarshubham) [reply demos/ota/ota_demo_core_http/CMakeLists.txt]: changed cmake in https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1425/commits/9f18e9905580d6e4067eaa4a215da9dbe1e2453a

**Review by tgsong — APPROVED** [review]

**Review by cobusve — APPROVED** [review]

**Comment by abhidixi11** [comment]
Merged , unit tests coverage will be fixed later.

**Pull request merged**